File: LeetCode/c++/BestTimeToBuyAndSellStockII.cpp
#include <iostream>
#include <vector>

using namespace std;

int maxProfit(vector<int>& prices) {
	if (prices.size() <= 1)
		return 0;

	int max = 0;

	//遍历股票价格，寻找增长最快的利润
	for (int i = 1; i < prices.size(); i++)
	{
		if (prices[i] > prices[i - 1]) //只要增长就增加
			max += prices[i] - prices[i - 1];
	}

	return max;
}

int main()
{
	vector<int> v;

	//max profit:8
	for (int i = 1; i < 10; i++)
		v.push_back(i);

	cout << maxProfit(v) << endl;

	//max profit:2
	v.clear();
	v.push_back(2);
	v.push_back(1);
	v.push_back(2);
	v.push_back(0);
	v.push_back(1);

	cout << maxProfit(v) << endl;

	//max profit:1
	v.clear();
	v.push_back(2);
	v.push_back(1);
	v.push_back(2);
	v.push_back(1);
	v.push_back(0);

	cout << maxProfit(v) << endl;

	//max profit:3
	v.clear();
	v.push_back(2);
	v.push_back(1);
	v.push_back(2);
	v.push_back(4);
	v.push_back(3);

	cout << maxProfit(v) << endl;

	//max profit:7
	v.clear();
	v.push_back(2);
	v.push_back(1);
	v.push_back(2);
	v.push_back(4);
	v.push_back(3);
	v.push_back(0);
	v.push_back(4);
	v.push_back(3);

	cout << maxProfit(v) << endl;

	//max profit:5
	v.clear();
	v.push_back(2);
	v.push_back(1);
	v.push_back(2);
	v.push_back(4);
	v.push_back(3);
	v.push_back(0);
	v.push_back(2);
	v.push_back(-1);

	cout << maxProfit(v) << endl;

	//max profit:7
	v.clear();
	v.push_back(6);
	v.push_back(1);
	v.push_back(3);
	v.push_back(2);
	v.push_back(4);
	v.push_back(7);
	v.push_back(2);
	v.push_back(-1);

	cout << maxProfit(v) << endl;

	return 0;
}

File: LeetCode/iter1/python/BestTimeToBuyAndSellStockII.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        listSize = len(prices)
        if(listSize<=1):
            return 0
			
        profitSum = 0
        for i in range(1,listSize,1):
            diff = prices[i]-prices[i-1]
            if(diff>0): #价格增长
                profitSum += diff
				
        return profitSum
        
#测试程序
s = Solution()
#max profit: 8
prices = [1,2,3,4,5,6,7,8,9]
print(s.maxProfit(prices))

#max profit: 2
prices = [2,1,2,0,1]
print(s.maxProfit(prices))

#max profit: 1
prices = [2,1,2,1,0]
print(s.maxProfit(prices))

#max profit: 3
prices = [2,1,2,4,1]
print(s.maxProfit(prices))

#max profit: 7
prices = [2,1,2,4,3,0,4,3]
print(s.maxProfit(prices))

#max profit: 5
prices = [2,1,2,4,3,0,2,-1]
print(s.maxProfit(prices))

#max profit: 7
prices = [6,1,3,2,4,7,2,-1]
print(s.maxProfit(prices))